tests/TrapTests.cs

```diff
@@ -46,6 +46,56 @@ public void ItIncludesAStackTrace()
                 .WithMessage("wasm trap: wasm `unreachable` instruction executed*");
         }
 
+        [Fact]
+        public void ItReturnsATrapCodeResult()
+        {
+
+            var instance = Linker.Instantiate(Store, Fixture.Module);
+            var run = instance.GetFunction<Result>("run_div_zero");
+            var result = run();
+
+            result.Type.Should().Be(ResultType.Trap);
+            result.Trap.Should().Be(TrapCode.IntegerDivisionByZero);
```

Comment:
Should we be asserting here (and in `ItReturnsATrapCodeGenericResult`) that the frames are empty?

Reply:
Unless I'm misunderstanding what you're asking for I don't think that's possible in those two cases. They're both `Result` types (not `ResultWithBacktrace`) so the only thing available is the `TrapCode`, the backtrace is not available.

Reply:
Oh right, the name `Trap` threw me off here. It should probably be, at the minimum, `TrapCode` so it doesn't seem like the whole trap exception is being stored.

Stepping back, though, should we not store a `TrapException` for the non-backtrace capturing `Result` types too, just without `Frames` populated?

I think it'd be useful to get the message in addition to the code, just without the backtrace.

Reply:
Good point on the naming, I'll change that now 👍

Regarding the `TrapException`/`message` my intention was to have the `Result` be as lightweight as possible, avoiding allocating a `TrapException` and the message string. Personally I would prefer to keep it that way.

Maybe the problem here is naming - `Result` (the simpler name) is the opt-in to a lighter weight version with no useful debugging information. Perhaps instead they should be named `Result` (with a backtrace) and `ResultWithoutAnyUsefulInformation` (pending a better name). That way it's probably more obvious that the version with the backtrace is the one you "should" use?

---

Having said all that, I'm wondering if another possible way to do this is instead of having some special "blessed types" which can interact with traps we should instead detect if the return type implements a special _interface_ and then do things through that. That way we can ship a single `Result` type with backtraces/messages etc, but there's a door to hook into the trap system (in a safe way of course) and get just what you need. If you don't mind holding off on this PR for a while I'll experiment with that approach to see if it's feasible.

Reply:
Sounds good, I'll hold off on the PR to see what you come up with.

I'm also good with landing this PR if that's what you decide is best.